test/sqllogictest/outer_join_simplification.slt

```diff
@@ -302,3 +307,553 @@ Source materialize.public.foo
+  filter=((#0) IS NOT NULL)
+
+EOF
+
+# Manually written idempotent semijoin.
+# Similar to the previous test, but the CTE from inside c0 is manually lifted to the enclosing LetRec.
+# The resulting plan should have 1 join with 2 inputs.
+query T multiline
+explain
+with mutually recursive
+  keys(a int4) as (
+    select distinct c0.a from c0, bar where c0.a = bar.a
+  ),
+  c0(a int4, b int8, u text, key int4) as (
+    select c0.a, c0.b, c0.u, c0.key from c0, keys where c0.a = keys.a
+    union
+    select * from c0
+  )
+select * from c0;
+----
+Explained Query:
+  Return
+    Get l0
+  With Mutually Recursive
+    cte l0 =
+      Distinct group_by=[#0..=#3]
+        Union
+          Project (#0..=#3)
+            Join on=(#0 = #4) type=differential
+              ArrangeBy keys=[[#0]]
+                Get l0
+              ArrangeBy keys=[[#0]]
+                Project (#0)
+                  Filter (#0) IS NOT NULL
+                    Get materialize.public.bar
+          Get l0
+
+Source materialize.public.bar
+  filter=((#0) IS NOT NULL)
+
+EOF
+
+# Manually written idempotent semijoin.
+# Similar to the previous test, but instead of using bar directly, we add an extra cte (bar2).
+# The resulting plan should have 1 join with 2 inputs.
+query T multiline
+explain
+with mutually recursive
+  bar2(a int4, v text) as (
+    select * from bar
+    union
+    select a, u from c0
+  ),
+  keys(a int4) as (
+    select distinct c0.a from c0, (select distinct on(a) * from bar2) as bar3 where c0.a = bar3.a
+  ),
+  c0(a int4, b int8, u text, key int4) as (
+    select c0.a, c0.b, c0.u, c0.key from c0, keys where c0.a = keys.a
+    union
+    select * from c0
+  )
+select * from c0;
+----
+Explained Query:
+  Return
+    Get l0
+  With Mutually Recursive
+    cte l0 =
+      Distinct group_by=[#0..=#3]
+        Union
+          Project (#0..=#3)
+            Join on=(#0 = #4) type=differential
+              ArrangeBy keys=[[#0]]
+                Get l0
+              ArrangeBy keys=[[#0]]
+                TopK group_by=[#0] limit=1 monotonic=false
+                  Project (#0)
+                    Distinct group_by=[#0, #1]
+                      Union
+                        Filter (#0) IS NOT NULL
+                          Get materialize.public.bar
+                        Project (#0, #2)
+                          Get l0
+          Get l0
+
+Source materialize.public.bar
+  filter=((#0) IS NOT NULL)
+
+EOF
+
+# Similar to the previous test, but SemijoinIdempotence should NOT kick in here! (That is, 2 joins should be in the
+# plan.) This is because we reordered the bindings compared to the previous test, and now the `bar2` reference in `keys`
+# changes meaning when `bar2` is being assigned after `key`. To correctly handle this situation, we need the expirations
+# in SemijoinIdempotence. To demonstrate this, if we comment out the two `do_expirations` lines, SemijoinIdempotence
+# incorrectly transforms this plan.
```

Comment:
:thumbsup: this is the type of test + code comment that _really_ helps with the review of the new code.

This and the above tests probably should be repeated as `*.spec` tests in #18714 because at the moment `NormalizeLets` will inline `keys` and `bar2` in the above query, so the expiration code is not exercised on the happy path (it would be if you reference `keys` and `bar2` in the `LetRec` body).

Reply:
I'll add these to the spec tests after they are merged.

If I changed the body in the happy test to refer to `bar2` and `keys`, then SemijoinIdempotence couldn't kick in, because then the intermediate join result is being referenced. I added this as another negative test.